astropy/io/registry/core.py

```diff
@@ -332,6 +339,12 @@ def write(self, data, *args, format=None, **kwargs):
 
             .. versionadded:: 4.3
         """
+        # Conservatively attempt to apply expanduser to the first argument,
```

Comment:
Since this is repeating the same code, you should factor this out to a little helper function like:
```
def _expand_user_in_args(args):
    <the same code>
    return args
```
Then:
```
# Expand user (~ or ~user) if applicable in args[0]
args = _expand_user_in_args(args)
```